Extract `prep_*` functions from `plot_*` functions

jdhoffa opened the pull request:
My plan here is to:

- [x] Extract, document and export the `prep_*` functions as-is, without touching the `plot_*` functions. Write tests for them, and ensure everything passes (just a refactor).
- [x] Alter the `plot_*` functions to expect the prepared data (also alter these expectations in the relevant tests)
- [x] Make sure to remove any unused `parameters` in the `plot_*` function
- [x] Complete deprecation process
- [x] Update `NEWS.md`
- [x] Review and check all `FIXME`s 

Closes #465

jdhoffa commented:
@MonikaFu I am working towards completing this, but I don't understand what the `env` argument for `prep_techmix` achieves. It seems to not be user-facing, but I'm curious how it should be handled. 

https://github.com/RMI-PACTA/r2dii.plot/blob/bc4814146bd0f8668b4d85c40a5b582774141a91/R/plot_techmix.R#L59-L70

MonikaFu commented:
You're right, it is not supposed to be user-facing. The variable is used in internal functions in order to return more meaningful errors. It passes the environment in which the function was called to the inside of the functions so that the name of variable that was used as for example, `data` argument is known and shown to the user in an error message. It should not be part of `prep_*` functions when they become user-facing.

jdhoffa commented:
Thanks for the context @MonikaFu , I will have a look at what you did, and come back if I have questions :-)

jdhoffa commented:
Ooof @MonikaFu this is a huge undertaking it turns out. 
I have done a good chunk of the extraction and refactoring (including updating tests) and I think we are at a functional step. 

I have 2 big TO-DOs left:

- [ ] Complete the deprecation process using the `lifecycle` package
- [ ]  Review and check all FIXMEs (I left two tests empty to test some behaviour that I wasn't sure how to test at the time. Going to come back to those in a moment)

But in the mean-time, when you're feeling better, I would love a quick review already, as a lot of files have been touched.

jdhoffa commented on NEWS.md:
```diff
@@ -1,6 +1,8 @@
 # r2dii.plot (development version)
 
 ## Breaking change
+* All `plot_*()` functions have had the data preparation step extracted as 
+`prep_*()` functions (#465).
```
Feel free to re-word this. I wonder if it maybe deserves more explanation in the `NEWS.md` as it is a Breaking and user-facing API change.

jdhoffa replied:
@MonikaFu to reword this if you like!

jdhoffa commented on R/plot_emission_intensity.R:
```diff
@@ -1,95 +1,22 @@
-#' @template convert_label
-#' @templateVar fun qplot_emission_intensity
-#' @templateVar value to_title
+#' @param data A data frame like the output of `prep_emission_intensity()`.
```
Most `plot_*` functions don't have that many arguments, and most get passed to the `prep_*` functions (you will see a similar trend with the tests.

jdhoffa commented on R/plot_emission_intensity.R:
```diff
@@ -1,95 +1,22 @@
 #'
 #' @export
 #' @examples
-#' # `data` must meet documented "Requirements"
-#' data <- subset(sda, sector == "cement" & region == "global")
-#' plot_emission_intensity(data)
-#'
 #' # plot with `qplot_emission_intensity()` parameters
-#' plot_emission_intensity(
-#'   data,
-#'   span_5yr = TRUE,
-#'   convert_label = to_title
-#' )
```
Examples had to be hugely refactored. To recreate the `qplot_*` functionality, the arguments have to be passed to the corresponding `prep_*` function. I still thought the info was useful, so I made that the ONLY example. 

I can add more examples if you think it is useful.

jdhoffa commented on R/plot_emission_intensity.R:
```diff
@@ -1,95 +1,22 @@
-plot_emission_intensity <- function(data,
-                                    span_5yr = FALSE,
-                                    convert_label = identity) {
-  lifecycle::deprecate_soft(
-    when = "0.4.0",
-    what = "plot_emission_intensity(data = 'must be prepped already')",
-    details = api_warning_details(
-      "prep_emission_intensity",
-      "plot_emission_intensity"
-      )
-  )
```
I have removed the deprecation warnings entirely. I have a TODO to re-add an appropriate message (I think? do I even need to?? I need to look into how to do this properly).

MonikaFu replied:
I'd say it would be great if we could add an informative error message in case the user calls plot on unprepped data.  Also - maybe we should write a blog post explaining how the functions work now and why we introduced this change? I guess I could write it if needed.

jdhoffa replied:
We don't have a blog currently :-D lol

jdhoffa replied:
I wonder if this should also be handled in #548 ?

MonikaFu replied:
Yes, I agree, it should be part of the checks in #548 .

jdhoffa commented on R/plot_emission_intensity.R:
```diff
@@ -1,95 +1,22 @@
-  specs <- dplyr::bind_cols(metrics, colours)
-
-  left_join(out, specs, by = metric(data))
+  check_emission_intensity(data, env = env)
```
`check_plot_emission_intensity` and `check_prep_emission_intensity` ended up being identical, so I just refactored it out entirely.

I do think it's good to do checks on both functions, since they are both now exported.

jdhoffa replied:
I agree in principle that the `checks` likely should be different as they test different input. However in practice, the were checking the presence of the exact same columns. 

That to me suggests that `check_plot_emission_intensity` and `check_prep_emission_intensity` likely should be expanded to include more precise checking, but I think that is out of scope for this PR. 

I will open a new issue to that effect.

jdhoffa replied:
Would you like me to split them back into `check_plot_emission_intensity` and `check_prep_emission_intensity` here? Recognizing that they would just be wrapping the same checks?

MonikaFu replied:
They used to check for the same things because they were both using unprepped data. But now since `plot` requires prep data I'd expect that it's checks should change and check if the data passed to it indeed is prepped.

jdhoffa replied:
Will be handled in #548

jdhoffa commented on R/plot_techmix.R:
```diff
@@ -226,22 +151,6 @@ get_technology_colours <- function(data) {
     )
 }
 
-recode_sector <- function(x) {
-  # styler: off
-  case_when(
-    grepl("(?i)power(?-i)", x)             ~ "power",
-    grepl("(?i)auto(?-i)[a-zA-Z]+", x)     ~ "automotive",
-    grepl("(?i)oil(?-i).*(?i)gas(?-i)", x) ~ "oil&gas",
-    grepl("(?i)fossil(?-i)[a-zA-Z]+", x)   ~ "fossil fuels",
-    TRUE ~ tolower(x)
-  )
-  # styler: on
-}
-
-extract_scenarios <- function(x) {
-  unique(x[startsWith(x, "target_")])
-}
-
```
I was refactoring to ensure all functions that were relevant to `prep_*` or `plot_*` were in the appropriate place.
While doing this, I noticed that these are used by multiple functions so i moved them to `utils.R`

jdhoffa commented on R/plot_trajectory.R:
```diff
@@ -268,57 +205,6 @@ order_trajectory <- function(data) {
-  lower <- min(data$value, na.rm = TRUE)
-  upper <- max(data$value, na.rm = TRUE)
-  span <- upper - lower
-  lower <- lower - 0.1 * span
-  upper <- upper + 0.1 * span
-
-  upper_distance <- distance_from_start_value(data, upper) / span
-  lower_distance <- distance_from_start_value(data, lower) / span
-
-  if (center_y) {
-    # Center the starting point of the lines
-    distance <- abs(upper_distance - lower_distance)
-    max_distance <- 0.1
-    if (distance > max_distance) {
-      lower <- lower - max(0, upper_distance - lower_distance) * span
-      upper <- upper + max(0, lower_distance - lower_distance) * span
-    }
-  }
-  list(lower = lower, upper = upper)
-}
-
-get_ordered_scenarios <- function(data) {
-  ordered_scenarios <- data %>%
-    filter(is_scenario(.data$metric), .data$year == max(data$year)) %>%
-    arrange(desc(.data$value)) %>%
-    pull(.data$metric) %>%
-    as.character()
-
-  ordered_scenarios
-}
-
-get_tech_kind <- function(data) {
-  technology_kind <- r2dii.data::increasing_or_decreasing %>%
-    filter(.data$technology == unique(data$technology)) %>%
-    pull(.data$increasing_or_decreasing) %>%
-    unique()
-
-  technology_kind
-}
-
```
All of these are now relevant to `prep_trajectory`

jdhoffa replied:
I believe this is done

jdhoffa commented on R/plot_trajectory.R:
```diff
@@ -352,67 +238,10 @@ get_ordered_scenario_colours <- function(n) {
-  data_worse_than_scenarios <- tibble(
-    year = unique(data$year),
-    technology = unique(data$technology),
-    sector = unique(data$sector)
+         "2" = pick(c("light_green", "red")),
+         "3" = pick(c("light_green", "light_yellow", "red")),
+         "4" = pick(c("light_green", "dark_yellow", "light_yellow", "red")),
+         "5" = scenario_colours,
+         abort(c("`n` must be between 2 and 5.", x = glue("Provided: {n}."))) # nocov
   )
```
git diff is getting confused here, this was a pure refactor

jdhoffa commented on R/plot_trajectory.R:
```diff
@@ -352,67 +238,10 @@ get_ordered_scenario_colours <- function(n) {
-    data_worse_than_scenarios$value <- area_borders$lower
-    data_worse_than_scenarios$metric <- "target_worse"
-    data_worse_than_scenarios$label <- "target_worse"
-
-    data_scenarios <- bind_rows(data_scenarios, data_worse_than_scenarios)
-
-    data_scenarios <- data_scenarios %>%
-      group_by(.data$year, .data$technology, .data$sector) %>%
-      mutate(metric = factor(.data$metric,
-        levels = rev(get_ordered_scenarios(data_scenarios))
-      )) %>%
-      arrange(.data$year, .data$metric) %>%
-      rename(value_low = "value") %>%
-      mutate(value = lead(.data$value_low,
-        n = 1,
-        default = area_borders$upper
-      ))
-  } else {
-    data_worse_than_scenarios$value <- area_borders$upper
-    data_worse_than_scenarios$metric <- "target_worse"
-    data_worse_than_scenarios$label <- "target_worse"
-
-    data_scenarios <- data %>%
-      filter(is_scenario(.data$metric))
-
-    data_scenarios <- bind_rows(data_scenarios, data_worse_than_scenarios)
-
-    data_scenarios <- data_scenarios %>%
-      group_by(.data$year, .data$technology, .data$sector) %>%
-      mutate(
-        metric = factor(
-          .data$metric,
-          levels = rev(get_ordered_scenarios(data_scenarios))
-        )
-      ) %>%
-      arrange(.data$year, .data$metric) %>%
-      mutate(value_low = lag(.data$value, n = 1, default = area_borders$lower))
-  }
-
-  data_scenarios
```
This logic was left alone, and purely refactored to `prep_trajectory`

jdhoffa replied:
I believe this is done

jdhoffa commented on R/qplot_emission_intensity.R:
```diff
@@ -18,7 +18,7 @@
 qplot_emission_intensity <- function(data) {
   env <- list(data = substitute(data))
-  check_plot_emission_intensity(data, env = env)
+  check_emission_intensity(data, env = env)
```
See above: pure refactor

jdhoffa commented on R/qplot_techmix.R:
```diff
@@ -30,8 +30,7 @@ qplot_techmix <- function(data) {
       convert_label = recode_metric_techmix,
       span_5yr = TRUE,
-      convert_tech_label = spell_out_technology,
-      env = env
```
`env` is no longer a user-facing argument of `prep_techmix`

jdhoffa commented on R/qplot_trajectory.R:
```diff
@@ -63,3 +63,17 @@ labs_trajectory <- function(p, data) {
     )
 }
+
+check_qplot_trajectory <- function(data, value_col, env) {
+  stopifnot(is.data.frame(data))
+  crucial <- c(common_crucial_market_share_columns(), value_col)
+  hint_if_missing_names(abort_if_missing_names(data, crucial), "market_share")
+  abort_if_has_zero_rows(data, env = env)
+  enforce_single_value <- c("sector", "technology", "region", "scenario_source")
+  abort_if_multiple(data, enforce_single_value, env = env)
+  abort_if_invalid_scenarios_number(data)
+  abort_if_too_many_lines(max = 4, summarise_max_year_by_scenario(data))
+  abort_if_too_many_lines(max = 5, summarise_max_year_by_traj_metric(data))
+
+  invisible(data)
+}
```
Added a new check, since `qplot_trajectory` now needs the check conditions for both `prep` and `plot`

jdhoffa replied:
NIT: I actually think that may be a bit of a dangerous assumption. I tend to think that any exported function should have it's own input checks. 

Consider the (unlikely) case that we decide to deprecate the `prep_*` and `plot_*` functions, and decide to remove all of the checks from them. `qplot_*` would not realize this, and would be left checkless.

But of course, you are maintainer, and if you would prefer I remove the checks here I am happy to do so!

jdhoffa replied:
Closing a resolved, see all the `qplot_` functions now :-)

jdhoffa commented on tests/testthat/test-plot_emission_intensity.R:
```diff
@@ -31,44 +31,18 @@ test_that("outputs an object with no factor-columns derived from `specs`", {
   expect_equal(metrics, ugly)
 })
 
-test_that("with too many lines to plot errors gracefully", {
-  data <- filter(sda, sector == "cement") %>%
-    bind_fake_sda_metrics(8)
-  expect_snapshot_error(plot_emission_intensity(data))
-})
-
-test_that("is sensitive to `convert_label`", {
-  data <- filter(sda, sector == "cement")
-
-  labels_def <- plot_emission_intensity(data) %>%
-    unique_plot_data("label")
-  labels_mod <- plot_emission_intensity(data, convert_label = toupper) %>%
-    unique_plot_data("label")
-
-  expect_false(identical(labels_def, labels_mod))
-})
-
-test_that("is sensitive to `span_5yr`", {
-  data <- filter(sda, sector == "cement")
-  abort_if_year_range_is_5yr_already(data)
-
-  p_f <- plot_emission_intensity(data, span_5yr = FALSE)
-  expect_false(diff(year_range(p_f)) == 5)
-
-  p_t <- plot_emission_intensity(data, span_5yr = TRUE)
-  expect_true(diff(year_range(p_t)) == 5)
-})
-
```
Many tests were moved to the more appropriate place (e.g. if "too many line" handling is in the `prep_*` function, it has been moved there.

MonikaFu replied:
We need to update the message or move the check I'd say. It is not very intuitive to have a message about too many lines to plot in a function that is not plot.

jdhoffa replied:
Ok so, to move this test to `plot_emission_intensity`, I had to update the checks (so that the lin check only occurs on plot, which totally makes sense). 

Doing so split the `check_*_emission_intensity` function up again, as there is now 1 additional check that occurs only for `plot_emission_intensity`. However, I will leave #548 open, as I imagine we should still add many checks  to prove that `plot_*` isn't accepting un-prepared data...

jdhoffa commented on tests/testthat/test-plot_emission_intensity.R:
```diff
@@ -80,14 +54,6 @@ test_that("with n metrics in input outputs n lines", {
 
 options(warn = 0)
 
-test_that("throws expected warning about API change", {
-  data <- head(filter(sda, sector == "cement"))
-  expect_snapshot_error(
-    plot_emission_intensity(data),
-    class = "warning"
-  )
-})
-
```
All "API change" warning tests have been removed. See my "figure out deprecation" TODO

jdhoffa commented on tests/testthat/test-plot_techmix.R:
```diff
@@ -4,49 +4,9 @@ test_that("without a `data` frame errors gracefully", {
-  bad_kind <- filter(sda, sector == first(sector))
-  expect_snapshot_error(plot_techmix(bad_kind))
-})
-
-test_that("with zero-row data errors gracefully", {
-  zero_row <- market_share[0L, ]
-  expect_snapshot_error(
-    plot_techmix(zero_row)
-  )
-})
-
-test_that("with more than one scenario errors gracefully", {
-  prep <- example_market_share()
-  expect_snapshot_error(plot_techmix(prep))
-})
-
-test_that("with too many sectors errors gracefully", {
-  bad_sector <- head(market_share, 2L)
-  bad_sector$sector <- c("a", "b")
-  expect_snapshot_error(plot_techmix(bad_sector))
-})
-
-test_that("with too many regions errors gracefully", {
-  bad_region <- head(market_share, 2L)
-  bad_region$region <- c("a", "b")
-  expect_snapshot_error(plot_techmix(bad_region))
-})
-
-test_that("with too many scenario_source errors gracefully", {
-  bad_scenario_source <- head(market_share, 2L)
-  bad_scenario_source$scenario_source <- c("a", "b")
-  expect_snapshot_error(plot_techmix(bad_scenario_source))
-})
-
-test_that("with too few scenarios errors gracefully", {
-  too_few <- head(market_share, 2L)
-  too_few$metric <- c(main_line(), "corporate_economy")
-  expect_snapshot_error(plot_techmix(too_few))
-})
-
```
Tests migrated to `prep`

MonikaFu replied:
potentially we might want to keep them in both? Not sure.

jdhoffa replied:
I think this should be handled in #548, when we add more checks to the`prep_*` and `plot_*` functions (we can also add tests that ensure each check functions as expected.

jdhoffa commented:
Regarding the deprecation process, my assessment is the following:
* We have adequately warned users that we plan to change the API of an existing function (a breaking change)
* Since we are not yet in version `1.0.0` of `r2dii.plot`, we have still allowed ourselves the opportunity to have breaking changes
* We will need to do a minor version bump for sure, as per SemVer
* I do not think we need any further deprecation messages (I think it's fine to simply remove them all)

jdhoffa commented on tests/testthat/test-plot_trajectory.R:
```diff
@@ -10,48 +10,51 @@ test_that("isn't restricted to plotting only 5 years", {
-  bad_region <- head(market_share, 2L)
-  bad_region$region <- c("a", "b")
-  expect_error(plot_trajectory(bad_region), "bad_region")
+  #FIXME: Figure out this test
+  # # Keep even if already tested in qplot_. Non-standard evaluation is fragile
+  # bad_region <- head(market_share, 2L) %>%
+  #   mutate(region = c("a", "b")) %>%
+  #   prep_trajectory()
+  #
+  # expect_error(plot_trajectory(bad_region), "bad_region")
```
I will need some support here. I seem to have broken this test (it no longer errors as expected), and I am sure it has to do with the `.env` argument. Would love to pair with you for ~20min to try to figure this bit out :-)

MonikaFu replied:
let's

jdhoffa commented:
There is a NOTE when running R CMD check:
"Namespace in Imports field not imported from: ‘forcats’ All declared Imports should be used."

I think this note isn't related to this PR, and should be tackled in a separate issue/ PR.

MonikaFu commented on tests/testthat/_snaps/plot_techmix.md:
```diff
@@ -1,61 +1,9 @@
-    ! `data` must have all the expected names.
-    x Missing names: metric, technology, technology_share.
-
-# with zero-row data errors gracefully
-
-    `zero_row` must have some rows.
-    x `zero_row` has zero rows.
-
-# with more than one scenario errors gracefully
-
-    `prep$metric` must have a single scenario not 3.
-    i Do you need to pick one scenario? E.g. pick 'target_cps' with: `subset(prep, metric %in% c('projected', 'corporate_economy', 'target_cps'))`.
-    x Provided: target_cps, target_sds, target_sps.
-
-# with too many sectors errors gracefully
-
-    `bad_sector` must have a single value of `sector`.
-    i Do you need to pick one value? E.g. pick 'a' with: `subset(bad_sector, sector == 'a')`.
-    x Provided: a, b.
-
-# with too many regions errors gracefully
-
-    `bad_region` must have a single value of `region`.
-    i Do you need to pick one value? E.g. pick 'a' with: `subset(bad_region, region == 'a')`.
-    x Provided: a, b.
```
These are still relevant I'd say. Or do we move it all to `prep_()` and add a check for if the data have been prepped instead?

MonikaFu commented on tests/testthat/_snaps/plot_techmix.md:
```diff
@@ -1,61 +1,9 @@
-    i Do you need to pick one value? E.g. pick 'a' with: `subset(bad_region, region == 'a')`.
-    x Provided: a, b.
-
-# with too many scenario_source errors gracefully
-
-    `bad_scenario_source` must have a single value of `scenario_source`.
-    i Do you need to pick one value? E.g. pick 'a' with: `subset(bad_scenario_source, scenario_source == 'a')`.
-    x Provided: a, b.
-
-# with too few scenarios errors gracefully
-
-    `too_few$metric` must have one scenario.
-    x It has none.
-
```
Still relevant I'd say?

MonikaFu commented on tests/testthat/_snaps/prep_emission_intensity.md:
```diff
@@ -0,0 +1,6 @@
+# with too many lines to plot errors gracefully
```
I'd say this is more of a plot check than a prep check still.

MonikaFu commented on tests/testthat/test-plot_techmix.R:
```diff
@@ -4,49 +4,9 @@ test_that("without a `data` frame errors gracefully", {
-  expect_snapshot_error(plot_techmix(bad_kind))
-})
-
-test_that("with zero-row data errors gracefully", {
```
I guess we should keep this test here as well? Someone might pass a zero row data to plot still.

jdhoffa replied:
I think we should handle this also in #548 when we add more checks (and corresponding tests)

MonikaFu reviewed (changes requested):
Thanks for your work! See my comments. Happy to discuss verbally as well.

MonikaFu commented:
@jdhoffa see my review on the draft. Also - given these changes we need to update the documentation on the package website. I'd say let's do it in a separate PR because this one is already huge. Also happy to do it myself. Should we open an issue for it?

jdhoffa commented on R/plot_emission_intensity.R:
```diff
@@ -1,95 +1,22 @@
-
-  left_join(out, specs, by = metric(data))
+  check_emission_intensity(data, env = env)
+  plot_emission_intensity_impl(data)
```
Un-related: This would also close #529 :-)

MonikaFu replied:
yup! Two birds with one stone or something

jdhoffa commented on R/plot_trajectory.R:
```diff
@@ -26,57 +14,32 @@
 plot_trajectory <- function(data,
-                            span_5yr = FALSE,
-                            convert_label = identity,
-                            center_y = FALSE,
```
Got it!

jdhoffa commented:
Probably closes #529

jdhoffa commented:
There is a R CMD CHECK note that I believe is unrelated to this PR. Relates to #547

MonikaFu commented on R/qplot_emission_intensity.R:
```diff
@@ -18,11 +18,18 @@
 #' qplot_emission_intensity(data)
 qplot_emission_intensity <- function(data) {
   env <- list(data = substitute(data))
+  check_prep_emission_intensity(data, env = env)
```
just to confirm - you are calling the checks here (there are also inside `prep_emission_intensity` in order to be able to pass the right `env` to it? Maybe long term it should be handled in a more elegant way. Now we are calling the checks twice.

jdhoffa replied:
Correct! 
NIT: For safety, I do believe calling input checks explicitly on any exported function is prudent (if inelegant), for reasons stated in a different comment. But leave it to you to manage as you see fit :-)

MonikaFu commented on R/plot_trajectory.R:
```diff
@@ -26,94 +18,44 @@
 #'   perc_y_scale = TRUE
 #' )
 plot_trajectory <- function(data,
                             span_5yr = FALSE,
```
`span_5yr` should be part of `prep_trajectory`. The same as for `emission_intensity` and `techmix`.

jdhoffa replied:
Done!

jdhoffa replied:
Done

MonikaFu commented:
@jdhoffa I see checks failing also due to changes in `r2dii.analysis`. It seems like there was a name change of some of output columns. We should not solve it in this PR but it should be solved as the demo datasets in `r2dii.plot` should follow the output of `r2dii.analysis`.

jdhoffa commented:
> @jdhoffa I see checks failing also due to changes in `r2dii.analysis`. It seems like there was a name change of some of output columns. We should not solve it in this PR but it should be solved as the demo datasets in `r2dii.plot` should follow the output of `r2dii.analysis`.

That is unexpected to me! I will look into it in `r2dii.analysis` and get back to you.

jdhoffa commented:
> @jdhoffa I see checks failing also due to changes in `r2dii.analysis`. It seems like there was a name change of some of output columns. We should not solve it in this PR but it should be solved as the demo datasets in `r2dii.plot` should follow the output of `r2dii.analysis`.

tl;dr: Broken checks are unrelated to this PR.

In case you are interested:

The failed checks are coming from a "breaking change" in `r2dii.data`, where we have changed the expected format of the NACE codes (cc @jacobvjk)

* The breaking change happened here (where we changed how we format NACE codes), this introduced the bug: https://github.com/RMI-PACTA/r2dii.data/pull/327
* This PR should fix it, by aligning `loanbook_demo` with the new `NACE` code: https://github.com/RMI-PACTA/r2dii.data/pull/338

jdhoffa commented:
I will wait until https://github.com/RMI-PACTA/r2dii.data/pull/338 is merged before re-requesting a review, so that we make sure CI/CD works well.

jdhoffa commented:
All checks now passing, this is ready for review!

MonikaFu commented on NEWS.md:
```diff
@@ -1,6 +1,8 @@
 # r2dii.plot (development version)
 
 ## Breaking change
+* All `plot_*()` functions have had the data preparation step extracted into 
+`prep_*()` functions (#465).
```
```suggestion
* All `plot_*()` functions have had the data preparation step extracted into 
`prep_*()` functions. This means that from now on `prep_*()` function needs
to be called on `data` prior to `plot_*()`. The APIs of `qplot_*()` functions stay 
unchanged (#465).
```

MonikaFu reviewed (approved):
LGTM. See my proposed change to `NEWS.md`.

MonikaFu reviewed (approved):
LGTM

jdhoffa commented:
Going to wait for confirmation that this won't affect ongoing projects, then I will merge :-)

jdhoffa commented:
Confirmed by @daisy-pacheco and @jacobvjk
Merging!
